### ISSUE TITLE
Changes to Intel SGX DCAP default QPL collateral version

### ISSUE THREAD

**Issue by daniel-weisse** [openenclave/openenclave#4268]
With release 1.12 of Intel's DCAP package there were changes to `sgx_ql_qve_collateral_t` which is used for `sgx_ql_get_quote_verification_collateral()`.

Previous [version 1.11](https://download.01.org/intel-sgx/sgx-dcap/1.11/linux/docs/Intel_SGX_ECDSA_QuoteLibReference_DCAP_API.pdf#[{%22num%22%3A56%2C%22gen%22%3A0}%2C{%22name%22%3A%22XYZ%22}%2C69%2C177%2C0]):
```C
typedef struct _sgx_ql_qve_collateral_t {
   uint32_t version;        ///< For V1 and V2 APIs, the ‘version’ field with
                            ///< have a value of 1.  For V3 APIs, the
                            ///< ‘version’ field will have the value of 3.
//...
//...
} sgx_ql_qve_collateral_t;
```

Current [version 1.12](https://download.01.org/intel-sgx/sgx-dcap/1.12/linux/docs/Intel_SGX_ECDSA_QuoteLibReference_DCAP_API.pdf#%5B%7B%22num%22%3A56%2C%22gen%22%3A0%7D%2C%7B%22name%22%3A%22XYZ%22%7D%2C69%2C177%2C0%5D):
```C
typedef struct _sgx_ql_qve_collateral_t {
    uint8_t version[2];     ///< version[0] = major_version, version[1] =
                            ///< minor_version.  For PCS V1 and V2 APIs, the
                            ///< major_version = 1 and minor_version = 0 and
                            ///< the CRLs will be formatted in PEM. For
                            ///< PCS V3 APIs, the major_version = 3 and the
                            ///< minor_version can be either 0 or 1.  A
                            ///< minor_verion of 0 indicates the CRL’s are
                            ///< formatted in Base16 encoded DER.  A minor
                            ///< version of 1 indicates the CRL’s are
                            ///< formatted in raw binary DER.
//...
//...
} sgx_ql_qve_collateral_t;
```

Using Intel's default QPL, e.g. by following the [non ACC machine setup guide](https://github.com/openenclave/openenclave/blob/master/docs/GettingStartedDocs/Contributors/NonAccMachineSGXLinuxGettingStarted.md#3-set-up-intel-dcap-quote-provider-library-qpl)  to install it , and configuring the QPL to use API 3.1 `oe_get_sgx_quote_verification_collateral` raises `Invalid collateral version 65539`.

`/etc/sgx_default_qcnl.conf` configuration:
```
# PCCS server address
PCCS_URL=https://localhost:8081/sgx/certification/v3/

# To accept insecure HTTPS certificate, set this option to FALSE
USE_SECURE_CERT=FALSE

# You can use Intel PCS to get quote verification collateral 
#COLLATERAL_SERVICE=https://api.trustedservices.intel.com/sgx/certification/v3/

# If you use PCCS service to get quote verification collateral, you can specify which API version is to be used
# The legacy 3.0 API will return CRLs in HEX encoded DER format, while the new 3.1 API will return raw DER format
PCCS_API_VERSION=3.1

# Maximum retry times for QCNL. If RETRY is not defined or set to 0, no retry will be performed.
# It will first wait one second and then for all forthcoming retries it will double the waiting time
# By using RETRY_DELAY you disable this exponential backoff algorithm
#RETRY_TIMES=6

# Sleep this amount of seconds before each retry when a transfer has failed with a transient error
#RETRY_DELAY=10
```

OpenEnclave sees a version it is not familiar with (`collateral->version` is neither `SGX_QL_QVE_COLLATERAL_VERSION_1` nor `SGX_QL_QVE_COLLATERAL_VERSION_3`).
From `host/sgx/sgxquoteprovider.c`:
```C
if (collateral->version != SGX_QL_QVE_COLLATERAL_VERSION_1 &&
        collateral->version != SGX_QL_QVE_COLLATERAL_VERSION_3)
    {
        OE_RAISE_MSG(
            OE_INVALID_ENDORSEMENT,
            "Invalid collateral version %d",
            collateral->version);
    }
```
This raises `Invalid collateral version 65539`

The error in this instance can probably be fixed by checking for an additional version 3.1 
```C
#define SGX_QL_QVE_COLLATERAL_VERSION_3_1 (65539)
if (collateral->version != SGX_QL_QVE_COLLATERAL_VERSION_1 &&
        collateral->version != SGX_QL_QVE_COLLATERAL_VERSION_3 &&
        collateral->version != SGX_QL_QVE_COLLATERAL_VERSION_3_1)
```

**Comment by yentsanglee** [comment]
I see that [sgx_ql_lib_common.h](https://github.com/intel/SGXDataCenterAttestationPrimitives/blob/65eb9c9baa9d8952f94d0a177ceaa27915bacc4b/QuoteGeneration/quote_wrapper/common/inc/sgx_ql_lib_common.h#L179) still has this definition:

````
typedef struct _sgx_ql_qve_collateral_t
{
    uint32_t version;                      /// The version of PCCS/PCS API used to retrieve verification collateral.
                                           /// For V1 and V2 APIs, the ‘version’ field with have a value of 1.
                                           /// For V3 APIs, the ‘version’ field will have the value of 3
    ....
} sgx_ql_qve_collateral_t;
````

Changing version from `uint32_t` to `uint8 [2]` is not backward compatible. It completely changes the struct field offsets.

As mentioned, OE allows `SGX_QL_QVE_COLLATERAL_VERSION_3` but it expects RAW buffer (v3.1), not hex string (v3.0). I'd suggest OE to skip hex string support. I.e.

````
#define SGX_QL_QVE_COLLATERAL_VERSION_3_1 (65539)
if (collateral->version != SGX_QL_QVE_COLLATERAL_VERSION_1 &&
        collateral->version != SGX_QL_QVE_COLLATERAL_VERSION_3_1)
````

**Comment by yentsanglee** [comment]
Spec 1.12.1 is [here](https://download.01.org/intel-sgx/sgx-dcap/1.12.1/linux/docs/Intel_SGX_ECDSA_QuoteLibReference_DCAP_API.pdf#%5B%7B%22num%22%3A56%2C%22gen%22%3A0%7D%2C%7B%22name%22%3A%22XYZ%22%7D%2C69%2C177%2C0%5D)

```
union {
    uint32_t version; ///<‘version’ is the backward compatible legacy representation
    struct {
        uint16_t major_version; ///< For PCS V1 and V2 APIs, the
        uint16_t minor_version; ///< major_version = 1 and minor_version = 0
    } ///< the CRLs will be formatted in PEM. For
 };   ///< PCS V3 APIs, the major_version = 3 and the
      ///< minor_version can be either 0 or 1. A
      ///< minor_verion of 0 indicates the CRL’s are
      ///< formatted in Base16 encoded DER. A minor
      ///< version of 1 indicates the CRL’s are
      ///< formatted in raw binary DER.
```

**Comment by RRathna** [comment]
This issue is fixed by #4432.
